taichi/common/json_serde.h

```diff
@@ -76,8 +76,63 @@ struct FieldNameList {
   }
 };
 
+namespace type {
+template <typename T>
+using remove_cvref =
+    typename std::remove_cv<typename std::remove_reference<T>::type>;
+
+template <typename T>
+using remove_cvref_t = typename remove_cvref<T>::type;
+}  // namespace type
+
+template <typename T>
+struct has_ptr_serde {
+  template <typename T_>
+  static constexpr auto helper(T_ *) -> std::is_same<
```

Comment:
I don't know if it has a name. I borrowed it from `serialization.h`. @k-ye once made a lecture about it. There's a document on feishu about it.

Reply:
Trailing return types

But since C++ 11 this is a legal way to declare functions: 
```auto func(...) -> ret_type``` i.e. consistent with lambda expr

In this case `auto` is kind of like `fn` in rust...